src/DDTrace/Util/HexConversion.php

```diff
@@ -6,6 +6,29 @@ final class HexConversion
 {
     public static function idToHex($id)
     {
-        return str_pad(dechex($id), 16, "0", STR_PAD_LEFT);
+        $hex = dd_trace_dec_hex((string) $id);
+        return str_pad($hex, 16, "0", STR_PAD_LEFT);
+    }
+
+    public static function hexToInt($hex)
+    {
+        if (strlen($hex) > 16) {
+            throw new \InvalidArgumentException(
+                '\HexConversion::hexToInt() arg must be hexstring with <= 16 characters'
+            );
+        }
+
+        if (strlen($hex) == 16 && strpos("89abcdef", $hex[0]) !== false) {
```

Comment:
What's the magic string "89abcdef"?

Reply:
Anything over `7fffffffffffffff` is not going to have a 64bit signed representation so needs to be shifted right a bit per the original functionality of the id generator: https://github.com/signalfx/signalfx-php-tracing/blob/1c349b93875c6ff6362b24d7509539883cd9ac95/src/ext/random.c#L20

This is for backwards compatibility to report to the dd agent, which requires an integer in json encoding.

Reply:
@tigrannajaryan, if the first character of a 16 char long hex string is any of 8-f, then it has exceeded the maximum signed integer and `hexdec()` will not return an integer (coerces to float: https://www.php.net/manual/en/function.hexdec.php).  If this is the case, we bitshift right like the original id generator did to ensure the encoded integer is positive.

Reply:
updated to be more informative